### PR TITLE
finish task 4

### PR DESCRIPTION
**Pull request by li-wang-uk**

### CODE REVIEW

File: __tests__/app.test.js [hunk 1]
```diff
@@ -41,4 +41,45 @@ describe( "/api/", ()=> {
             expect(body.allEndPoints).toEqual(endpoints)
             })
         })
-    })
+})
+
+describe( "/api/articles/:article_id", ()=> {
+    it("GET 200 sends the selected article by given article_id", () => {
+        return request(app)
+        .get("/api/articles/1")
```
Comment (ghost) [inline __tests__/app.test.js:59]: You are requesting article 1, but expecting literally anything. How does this test give me and you confidence that your query is correct?

File: api/models/topics-models.js [hunk 2]
```diff
@@ -6,4 +6,22 @@ exports.selectTopics = () => {
...
     })
-}
\ No newline at end of file
+}
+
+exports.selectArticlesById = (article_id) => {
+    let queryString = `
+    SELECT * FROM articles
+    `
+    if (!Number(article_id)){
```
Comment (ghost) [inline api/models/topics-models.js]: PSQL will handle this, you do not need to promise reject. The database is specifically looking for a number, if you do not provide one you will get a 22P02 error which you should be handling.

File: api/models/topics-models.js [hunk 3]
```diff
@@ -6,4 +6,22 @@ exports.selectTopics = () => {
...
+        return Promise.reject({ status: 400, msg: "Bad Request" });
+    }
+    queryString += `WHERE article_id = ${article_id} ;`
```
Comment (ghost) [inline api/models/topics-models.js]: This should absolutely not be in template literals, this leaves you open for SQL injection.

**Review by ghost — CHANGES_REQUESTED** [review]
I need you to look at this again, your model is leaving you open to vulnerabilities and your test is not specific enough. Please make the changes and resubmit your PR.

File: __tests__/app.test.js [hunk 4]
```diff
@@ -44,26 +44,29 @@ describe( "/api/", ()=> {
...
             const selectedArticle = body.articles[0];
-            expect(typeof selectedArticle.article_id).toBe("number")
-            expect(typeof selectedArticle.author).toBe("string")
-            expect(typeof selectedArticle.title).toBe("string")
-            expect(typeof selectedArticle.article_id).toBe("number")
-            expect(typeof selectedArticle.body).toBe("string")
-            expect(typeof selectedArticle.created_at).toBe("string")
-            expect(typeof selectedArticle.votes).toBe("number")
-            expect(typeof selectedArticle.article_img_url).toBe("string")
+            expect(selectedArticle).toEqual(articleOneCopy)
```
Comment (nc-mick-fay) [inline __tests__/app.test.js:64]: Good change to this testing to be more specific in your expected response

File: __tests__/app.test.js [hunk 5]
```diff
@@ -72,9 +75,9 @@ describe( "/api/articles/:article_id", ()=> {
               });
           });
 
-    it("GET 404 when given article_id is not found", () => {
+    it("GET 404 when given article_id is not found in database", () => {
```
Comment (nc-mick-fay) [inline __tests__/app.test.js:78]: These small changes to test descriptions make it much easier to understand what you are testing for, good work!

File: api/errors.js [hunk 6]
```diff
@@ -5,7 +5,19 @@ exports.handleCustomErrors = (err, req, res, next) => {
...
+    if(err.code === '22P02'){
+        res.status(400).send({ msg: 'Bad Request' });
+    } else {
+        next(err);
+    }
+}
+
+exports.handleServerErrors = (err, req, res, next) => {
+    console.log(err)
+    res.status(500).send({ msg: 'Internal Server Error' });
```
Comment (nc-mick-fay) [inline api/errors.js:18]: Good additions of psql error handling and a general server error handler

File: api/models/topics-models.js [hunk 7]
```diff
@@ -10,14 +10,11 @@ exports.selectTopics = () => {
...
+    SELECT * FROM articles 
+    WHERE article_id = $1;
     `
-    if (!Number(article_id)){
-        return Promise.reject({ status: 400, msg: "Bad Request" });
-    }
-    queryString += `WHERE article_id = ${article_id} ;`
     return db
-    .query(queryString)
+    .query(queryString, [article_id])
```
Comment (nc-mick-fay) [inline api/models/topics-models.js:18]: Good parameterisation of your query to avoid SQL injection!

File: api/controllers/topics-controllers.js [hunk 8]
```diff
@@ -9,9 +9,8 @@ exports.getTopics = (req,res,next) => {
     .catch(next)
 }
 
-exports.getEndpoints = (req,res,next) => {
+exports.getEndpoints = (req,res) => {
     res.status(200).send({allEndPoints})
-    .catch(next)
```
Comment (nc-mick-fay) [inline api/controllers/topics-controllers.js:13]: As you're sending back a pre-built JSON object, I suppose a catch block isn't strictly necessary here

**Review by nc-mick-fay — APPROVED** [review]
Much improved test descriptions, and I can see positive changes in all requested areas

**Pull request merged**